### ISSUE TITLE
mupdf backend?

### ISSUE THREAD

**Issue by zabbal** [politza/pdf-tools#144]
There seems to lots of praise for http://mupdf.com/ libraries. I'm curious if you could glance over it and share your opinion whether this is worthy potential replacement to poppler.

**Comment by politza** [comment]
zabbal <notifications@github.com> writes:

> There seems to lots of praise for http://mupdf.com/ libraries.

Who praises it for what reasons ?

**Comment by vyp** [comment]
me for speed.

but not sure if mupdf provides all the same things that poppler does, which may be why it is so fast. doc-view mode can use mupdf as backend though, but for whatever reason I haven't found it as fast as simply using mupdf/zathura-pdf-mupdf.

**Comment by politza** [comment]
vyp <notifications@github.com> writes:

> me for speed.

Do you have a PDF where, let's say, okular is noticeably less performant ?

> but not sure if mupdf provides all the same things that poppler does

It looks pretty low-level, with no documentation at all in many areas.
I doubt there are any mupdf based viewers which support annotations ?

> doc-view mode can use mupdf as backend though, but for whatever reason
> I haven't found it as fast as simply using mupdf/zathura-pdf-mupdf.

doc-view renders everything to the file-system. Once that's done, the
back-end doesn't matter anymore.  After that, a lot of time is spent
uncompressing the PNG files inside Emacs.

**Comment by zakkak** [comment]
I don't know about speed, they both seam pretty good to me but I find poppler significantly more memory-hungry than mupdf.  I have compared with various viewers (evince, mupdf, zathura, pdf-tools) and found this behaviour consistent.

The [mupdf website](http://mupdf.com/) claims that:
> MuPDF is also small, fast, and yet complete. It supports PDF 1.7 with transparency, encryption, hyperlinks, **annotations**, searching and more. It also reads XPS and OpenXPS documents. MuPDF is written modularly, so features can be added on by integrators if they so desire.

**Comment by vyp** [comment]
@politza

> Do you have a PDF where, let's say, okular is noticeably less performant ?

http://projecteuclid.org/euclid.pjm/1102986512

> It looks pretty low-level, with no documentation at all in many areas.
I doubt there are any mupdf based viewers which support annotations ?

Yes I think you are correct. I don't think mupdf supports annotations.

> doc-view renders everything to the file-system. Once that's done, the
back-end doesn't matter anymore.  After that, a lot of time is spent
uncompressing the PNG files inside Emacs.

Does this package also do that?

**Comment by politza** [comment]
vyp <notifications@github.com> writes:

> [...] I don't think mupdf supports annotations.

It does, i.e. it's able to read the various PDF data-structures.  Though,
the high-level API (abstracting from the PDF standard details) looks
very much ad-hoc.

> Does this package also do that?

More or less.  The images are written to a temporary file and
immediately read back into a Lisp string.  The format is also PNG, but
with a low-compression rate, which makes Emacs display them faster.

Contrary to doc-view, images are created on demand. (Though there is some
speculative pre-loading involved.)

Finally, there is only one process involved.  I.e if you jump to a page
in doc-view (which has not been rendered yet), it starts a new process
for this page only.

**Comment by vyp** [comment]
@politza Ah I see. My point is that **if** you want to implement mupdf support, you're going to need some sort of advantage over just using mupdf backend with doc view mode. Because otherwise it would be useless and a waste of time.

So it could be something like possible speed improvements? If you're saying the bottleneck with doc view is decompressing png files in emacs, but you're also saying that pdf-tool's low compression rate pngs would be faster to display, that would be an improvement over doc view mode.

Otherwise any of the other features that you list in the "features" section in the readme, because, except for viewing, I don't think doc-view mode has that great support for a lot of the things that pdf-tools can do. (if i recall correctly, i haven't used doc-view mode in a while) Which is probably because doc-view mode can also show non-pdf documents too (I think).

In terms of mupdf features, when i use mupdf backend with zathura, I can search (not isearch like with pdf-tools, but it can highlight the search result(s) on the pdf, which doc view mode cannot do), and I can also use my mouse to highlight regions of text to copy them to the clipboard. So I imagine it's possible to at least do these things with the mupdf libraries (unless zathura itself is doing some very weird/complicated things).

@zakkak sorry I missed your comment. While I certainly find poppler based viewers usable, in general I find mupdf to be a lot faster in rendering them. Perhaps this is more to do with the pdfs I view though. But for example if I switch zathura's pdf backend from zathura-pdf-poppler to zathura-pdf-mupdf, it becomes faster at rendering. I generally don't view pdfs in emacs (with either doc view or pdf-tools) because it's too slow, curiously even slower than non-emacs poppler pdf viewers.

(Doc-view mode's mupdf backend is okay, except most of the time I need to view with slightly higher resolutions, which drastically reduces speed of rendering. For whatever reason zathura with mupdf backend can display comparable resolutions with basically no reduction in rendering speed, so I just use zathura.)

**Comment by politza** [comment]
vyp <notifications@github.com> writes:

> @politza Ah I see. My point is that if you want to implement mupdf support, you're going to need some sort of advantage
> over just using mupdf backend with doc view mode. 

No, that's pretty much decided.  What it needs to have, is an advantage
over poppler.

> Because otherwise it would be useless and a waste of time.

I certainly don't want to do that.

> So it could be something like possible speed improvements? 

That's the main thrust here.  It seems to render faster, at least in
some cases.

> If you're saying the bottleneck with doc view is decompressing png
> files in emacs,  [...]

Decompressing and loading into Emacs in general.

**Comment by vyp** [comment]
> No, that's pretty much decided.

As in, you have decided to give mupdf backend a go?

> What it needs to have, is an advantage over poppler.

I'm fairly certain it is faster in general than poppler, but only based on my empirical experiences.

Here are some links I found by searching "mupdf vs poppler" which back that up:

http://www.freesoftwaremagazine.com/articles/opening_large_pdf_files_mupdf_comes_rescue
https://github.com/xiangxw/mupdf-qt/wiki/Comparison-of-Mupdf-and-Other-PDF-Rendering-Libraries
http://stackoverflow.com/questions/7322768/is-mupdf-library-faster-than-xpdf-poppler-at-rendering-images-from-pdf-pages

Here's a post where the author has some actual numbers:

http://hzqtc.github.io/2012/04/poppler-vs-mupdf.html

But that was three years ago, and nor does author provide the exact method to replicate it (the pdf, the command/program).

The difference in speed is better seen with a greater number of pages. But does emacs itself have limits/difficulties displaying large (maybe 500/1000+ pages) pdfs? Does it for example require the entire pdf to be loaded into memory at once? Because if it does has some limits like that, then the speed advantages mupdf gains with larger pdfs become useless anyway with regard to using it for emacs.

That and as @zakkak says, less memory consumption. (although I haven't checked that myself) So it depends on whether you consider (potentially) greater speed and/or less memory consumption worthwhile advantages to justify spending time on this. (Remember, even though mupdf seems to be faster, maybe the scenario changes when using it for emacs somehow.)

**Comment by politza** [comment]
OK, mupdf renders faster than poppler (up to 4 times in my simplistic
comparison).

But mupdf seems to have problems with (at least) highlight annotations,
which are, in one case, rendered with a wrong color and misplaced
rectangle.  I expect there to be more bugs like this, since, at least
the annotation component isn't used widely (or is it ?) and no
test-cases exists at all.

Further, there is almost no documentation and the PDF interface is
incomplete.

In summary, while it would be nice to speed up the rendering, everything
else speaks against mupdf, so I don't think I want to use it.

BTW., in one case ~30% of render-time of poppler was spent in
libcms2, which is a library for color management.  I am not sure what
the advantage of using a CMS for the user is.  Maybe the poppler
people could be bugged to make this feature optional at run-time.

**Comment by vyp** [comment]
Fair enough. Thanks for taking the time to evaluate it. And I'll look into the usage of libcms2, thanks.
